### PR TITLE
[CI] fine-grained jobs

### PR DESCRIPTION
**Pull request by RingerJK**
## Description

CI: - fine-grained jobs: build(build project), unit tests, and static-analysis(lint , public API, and documentation checkers) + ui-robo-tests + ui-instrumentation-tests; - `requires` via workflow process

- [ ] I have added any issue links
- [x] I have added all related labels (`bug`, `feature`, `new API(s)`, `SEMVER`, etc.)
- [x] I have added the appropriate milestone and project boards

### Goal

Please describe the PR goals. Just the stuff needed to implement the fix / feature and a simple rationale. It could contain many check points if needed

### Implementation

Please include all the relevant things implemented and also rationale, clarifications / disclaimers etc. related to the approach used. It could be as self code companion comments

## Screenshots or Gifs

Please include all the media files to give some context about what's being implemented or fixed. It's not mandatory to upload screenshots or gifs, but for most of the cases it becomes really handy to get into the scope of the feature / bug being fixed and also it's REALLY useful for UI related PRs screenshot gif

## Testing

Please describe the manual tests that you ran to verify your changes

- [ ] I have tested locally (including `SNAPSHOT` upstream dependencies if needed) through testapp/demo app and run all activities to avoid regressions
- [ ] I have tested via a test drive, or a simulation/mock location app
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have updated the `CHANGELOG` including this PR
- [ ] We might need to update / push `api/current.txt` files after running `$> make core-update-api` (Core) / `$> make ui-update-api` (UI) if there are changes / errors we're 🆗 with (e.g. `AddedMethod` changes are marked as errors but don't break SemVer) 🚀 If there are SemVer breaking changes add the `SEMVER` label. See [Metalava](https://github.com/mapbox/mapbox-navigation-android/blob/master/docs/metalava.md) docs

### CODE REVIEW

File: circle.yml [hunk 1]
```diff
@@ -9,18 +9,37 @@ workflows:
...
       - build
+      - static-analysis:
+          requires:
+            - build
+      - unit-tests:
+          requires:
+            - build
+      - ui-robo-tests:
+          requires:
+            - build
```
Comment (LukasPaczos) [inline circle.yml:20]: This job runs `release` variants of libraries while `build` builds `debug` variants, so `ui-robo-tests` doesn't have to rely on `build`. How about we make `build` job a `build-debug` and add another `build-release` job that `ui-robo-tests`, `release-snapshot` and `release` can require?

File: circle.yml [hunk 2]
```diff
@@ -9,18 +9,37 @@ workflows:
...
+      - ui-intrumentation-tests:
+          requires:
+            - build
       - release-snapshot:
+          requires:
+            - build
```
Comment (LukasPaczos) [inline circle.yml:26]: Same as above.

File: circle.yml [hunk 3]
```diff
@@ -9,18 +9,37 @@ workflows:
...
+            - static-analysis
+            - unit-tests
           filters:
             branches:
               only:
                 - master
       - release:
+          requires:
+            - build
```
Comment (LukasPaczos) [inline circle.yml:35]: Same as above.

File: circle.yml [hunk 4]
```diff
@@ -290,17 +309,50 @@ jobs:
...
+      - build-ui-debug
+
+  unit-tests:
+    working_directory: ~/code
+    docker:
+      - image: << pipeline.parameters.docker-image-tag >>
+    steps:
+      - checkout
       - build-core-debug
       - build-ui-debug
```
Comment (LukasPaczos) [inline circle.yml:322]: This is already executed during `build` that this job requires.

File: circle.yml (hunk 1 above)
Comment (RingerJK) [reply circle.yml:20]: `build` combines all variants(debug+release), but I think we could replace it by `assemble`(that is assemble for _Release_ and _Debug_ at the same time), wdyt?

Comment (LukasPaczos) [reply circle.yml:20]: As far as I can see, `build` only generates debug artifacts - https://github.com/mapbox/mapbox-navigation-android/blob/e5ce9590acbd2ac7e2b079ac553af975fd98c561/circle.yml#L306-L314

Comment (RingerJK) [reply circle.yml:20]: @LukasPaczos nice note, I didn't expect that "build" is actually "assemble"
https://github.com/mapbox/mapbox-navigation-android/blob/e7be76c5af257195c87bef420be455d6a136fc96/Makefile#L66-L72
and 
https://github.com/mapbox/mapbox-navigation-android/blob/e7be76c5af257195c87bef420be455d6a136fc96/Makefile#L106-L112
what do you think if we rename these tasks at `make` file and rename tasks at `CI` config

**cc** @Guardiola31337

Comment (LukasPaczos) [reply circle.yml:20]: I don't mind, whatever works best :)

Comment (RingerJK) [reply circle.yml:20]: @LukasPaczos have suggestion: let's not run `firebase` tests(robo and instrumentation) if we don't archive all previous checks(assemble release+debug, unit tests, lint, docs, check api). 
Pros: 
- saves money(firebase is not free) 
- keeps more machines ready for other builds(we have limited range of machines at circle ci, correct me if I'm wrong)
- CI runs firebase tests anyway, when previous fails are solved
Cons:
- increase time of tests(ui and robo tests starts after all previous checks passed)

Comment (LukasPaczos) [reply circle.yml]: Sounds good @RingerJK :+1: 

@mapbox/navigation-android let's monitor the duration of our CI jobs and if they start getting out of hand, we can revisit this setup. For code-unrelated changes, we can work on https://github.com/mapbox/mapbox-navigation-android/issues/3544.

**Review by LukasPaczos — APPROVED** [review]

**Comment by RingerJK** [comment]
[Changed](https://github.com/mapbox/mapbox-navigation-android/settings/branch_protection_rules/4904782) `Require status checks to pass before merging`. Now required checks: 
- assemble-debug
- assemble-release
- static-analysis
- unit-tests

**cc** @mapbox/navigation-android

**Pull request merged**